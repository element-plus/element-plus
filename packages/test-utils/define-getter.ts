import { isFunction, isUndefined } from 'lodash-unified'

/**
 *
 * @param obj Object on which to add or modify the property.
 * @param prop The property name
 * @param value The value of `obj[prop]` or a getter
 * @returns A restore function which can reset `obj[prop]`'s value or getter
 */
const defineGetter = (
  obj: Record<string, any>,
  prop: string,
  value: any,
  defaultValue?: any
) => {
  let oldValue = defaultValue
  const { get, configurable } = Object.getOwnPropertyDescriptor(obj, prop) || {}

  if (isUndefined(defaultValue) && isUndefined(get)) {
    try {
      oldValue = obj[prop]
    } catch {
      throw new Error(
        `TypeError: Illegal invocation. Cannot read ${prop} of '${obj}', '${obj}' might be a prototype,  please specify default value instead.`
      )
    }
  }

  const oldGetter = get ?? (() => oldValue)
  const getter = isFunction(value) ? value : () => value

  Object.defineProperty(obj, prop, {
    configurable: true,
    get: getter,
  })

  return () => {
    Object.defineProperty(obj, prop, {
      configurable,
      get: oldGetter,
    })
  }
}

export default defineGetter
