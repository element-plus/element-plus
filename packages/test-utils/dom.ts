export const getCssVariable = (el: HTMLElement, property: string) => {
  return getComputedStyle(el).getPropertyValue(property)
}
