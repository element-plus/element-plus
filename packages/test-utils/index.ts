export * from './mock'
export * from './dom'
export { default as defineGetter } from './define-getter'
export { default as makeScroll } from './make-scroll'
export { default as sleep } from './sleep'
export { default as tick } from './tick'
export { default as triggerEvent } from './trigger-event'
export { default as stylePlugin } from './style-plugin'
export {
  default as triggerCompositeClick,
  triggerNativeCompositeClick,
} from './composite-click'
export { default as makeMount } from './make-mount'
