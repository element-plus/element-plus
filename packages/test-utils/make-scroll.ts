import sleep from './sleep'

const makeScroll = (
  dom: Element,
  name: 'scrollTop' | 'scrollLeft',
  offset: number
) => {
  const eventTarget = dom === document.documentElement ? window : dom
  dom[name] = offset
  const evt = new CustomEvent('scroll', {
    detail: {
      target: {
        [name]: offset,
      },
    },
  })
  eventTarget.dispatchEvent(evt)
  // must use setTimeout instead of nextTick to wait dom change
  return sleep()
}

export default makeScroll
