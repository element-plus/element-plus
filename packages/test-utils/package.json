{
  "name": "@element-plus/test-utils",
  "version": "0.0.5",
  "private": true,
  "license": "MIT",
  "main": "index.ts"
}
