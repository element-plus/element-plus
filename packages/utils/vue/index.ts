export * from './global-node'
export * from './icon'
export * from './install'
export * from './props'
export * from './refs'
export * from './size'
export * from './typescript'
export * from './validator'
export * from './vnode'
