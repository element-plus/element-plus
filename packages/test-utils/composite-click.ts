import { nextTick } from 'vue'
import triggerEvent from './trigger-event'

import type { DOMWrapper, VueWrapper } from '@vue/test-utils'

const triggerCompositeClick = async <
  T extends VueWrapper<any> | DOMWrapper<Element>
>(
  wrapper: T
) => {
  await wrapper.trigger('mousedown')
  await wrapper.trigger('mouseup')
  await wrapper.trigger('click')
}

export default triggerCompositeClick

export const triggerNativeCompositeClick = async (el: Element) => {
  triggerEvent(el, 'mousedown')
  await nextTick()
  triggerEvent(el, 'mouseup')
  await nextTick()
  triggerEvent(el, 'click')
  return nextTick()
}
