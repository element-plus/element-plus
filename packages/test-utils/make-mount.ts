import { mount } from '@vue/test-utils'
import { merge } from 'lodash-unified'

const makeMount = <C, O, E>(element: C, defaultOptions: O) => {
  return (props: (E | O) | (E & O) = {} as E) =>
    mount(element, merge({}, defaultOptions, props))
}

interface Options {
  data?: () => {
    [key: string]: unknown
  }
  methods?: {
    [key: string]: (...args: unknown[]) => any
  }
}

export const makeMountFunc = <T extends Record<string, unknown>>(
  defaultOptions: T
) => {
  return (template: string, options: Options) => {
    return mount({
      ...merge({}, defaultOptions, options),
      template,
    })
  }
}

export default makeMount
