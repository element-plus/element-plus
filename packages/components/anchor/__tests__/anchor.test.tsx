import { nextTick, ref } from 'vue'
import { mount } from '@vue/test-utils'
import { describe, expect, test, vi } from 'vitest'
import Anchor from '../src/anchor.vue'
import AnchorLink from '../src/anchor-link.vue'

let id = 0

const getHash = () => `#anchor-link-${id++}`

describe('Anchor.vue', () => {
  test('snapshot', async () => {
    const hash = getHash()
    const wrapper = mount({
      props: ['direction', 'type'],
      render() {
        return (
          <Anchor direction={this.direction} type={this.type}>
            <AnchorLink href={hash}>{hash}</AnchorLink>
          </Anchor>
        )
      },
    })

    wrapper.setProps({
      direction: 'vertical',
      type: 'default',
    })
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.setProps({
      direction: 'horizontal',
      type: 'default',
    })
    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.setProps({
      direction: 'vertical',
      type: 'underline',
    })
    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.setProps({
      direction: 'horizontal',
      type: 'underline',
    })
    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('scroll after click', async () => {
    const hash = getHash()
    const wrapper = mount(
      () => (
        <>
          <div id={hash.replace('#', '')}>hash</div>
          <Anchor>
            <AnchorLink href={hash}>{hash}</AnchorLink>
          </Anchor>
        </>
      ),
      {
        attachTo: 'body',
      }
    )
    const scrollSpy = vi
      .spyOn(window, 'scrollTo')
      .mockImplementation(() => undefined)
    await nextTick()
    wrapper.find(`a[href="${hash}"]`).trigger('click')
    await nextTick()
    expect(scrollSpy).toHaveBeenCalledTimes(1)
  })

  test('scrollTo method', async () => {
    const hash = getHash()
    const anchorRef = ref<InstanceType<typeof Anchor> | null>(null)
    mount(
      () => (
        <>
          <div id={hash.replace('#', '')}>hash</div>
          <Anchor ref={anchorRef}>
            <AnchorLink href={hash}>{hash}</AnchorLink>
          </Anchor>
        </>
      ),
      {
        attachTo: 'body',
      }
    )
    const scrollSpy = vi
      .spyOn(window, 'scrollTo')
      .mockImplementation(() => undefined)
    await nextTick()
    anchorRef.value?.scrollTo(hash)
    await nextTick()
    expect(scrollSpy).toHaveBeenCalledTimes(1)
  })

  test('change event', async () => {
    const hash1 = getHash()
    const hash2 = getHash()
    const wrapper = mount(() => (
      <Anchor>
        <AnchorLink href={hash1}>{hash1}</AnchorLink>
        <AnchorLink href={hash2}>{hash2}</AnchorLink>
      </Anchor>
    ))
    wrapper.find(`a[href="${hash1}"]`).trigger('click')
    wrapper.find(`a[href="${hash2}"]`).trigger('click')
    expect(wrapper.findComponent(Anchor).emitted('change')).toEqual([
      [hash1],
      [hash2],
    ])
  })

  test('click event', async () => {
    const hash = getHash()
    const wrapper = mount(() => (
      <Anchor>
        <AnchorLink href={hash}>{hash}</AnchorLink>
      </Anchor>
    ))
    wrapper.find(`a[href="${hash}"]`).trigger('click')

    expect(
      (wrapper.findComponent(Anchor).emitted('click') as any)[0][0]
    ).toBeInstanceOf(MouseEvent)
    expect((wrapper.findComponent(Anchor).emitted('click') as any)[0][1]).toBe(
      hash
    )
  })
})
