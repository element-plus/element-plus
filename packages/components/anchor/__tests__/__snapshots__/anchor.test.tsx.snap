// Vitest Snapshot v1

exports[`Anchor.vue > snapshot 1`] = `
"<div class=\\"el-anchor el-anchor--vertical\\">
  <div class=\\"el-anchor__marker\\"></div>
  <div class=\\"el-anchor__list\\">
    <div class=\\"el-anchor__item\\"><a class=\\"el-anchor__link\\" href=\\"#anchor-link-0\\">#anchor-link-0</a>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`Anchor.vue > snapshot 2`] = `
"<div class=\\"el-anchor el-anchor--horizontal\\">
  <div class=\\"el-anchor__marker\\"></div>
  <div class=\\"el-anchor__list\\">
    <div class=\\"el-anchor__item\\"><a class=\\"el-anchor__link\\" href=\\"#anchor-link-0\\">#anchor-link-0</a>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`Anchor.vue > snapshot 3`] = `
"<div class=\\"el-anchor el-anchor--underline el-anchor--vertical\\">
  <div class=\\"el-anchor__marker\\"></div>
  <div class=\\"el-anchor__list\\">
    <div class=\\"el-anchor__item\\"><a class=\\"el-anchor__link\\" href=\\"#anchor-link-0\\">#anchor-link-0</a>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`Anchor.vue > snapshot 4`] = `
"<div class=\\"el-anchor el-anchor--underline el-anchor--horizontal\\">
  <div class=\\"el-anchor__marker\\"></div>
  <div class=\\"el-anchor__list\\">
    <div class=\\"el-anchor__item\\"><a class=\\"el-anchor__link\\" href=\\"#anchor-link-0\\">#anchor-link-0</a>
      <!--v-if-->
    </div>
  </div>
</div>"
`;
